File: platforms/apple/Sources/BotaDeviceSDK/DeviceManager.swift
import BotaDeviceSDKC
import Foundation

public struct DeviceCapabilities: OptionSet, Equatable, Sendable {
    public let rawValue: UInt64

    public static let bluetooth = Self(rawValue: 1 << 0)
    public static let persistence = Self(rawValue: 1 << 1)
    public static let secureStorage = Self(rawValue: 1 << 2)
    public static let networkTransfer = Self(rawValue: 1 << 3)
    public static let recordingSink = Self(rawValue: 1 << 4)
    public static let firmwareBlob = Self(rawValue: 1 << 5)

    public init(rawValue: UInt64) { self.rawValue = rawValue }

    init(core: CoreCapabilities) {
        var value: Self = []
        if core.contains(.bluetooth) { value.insert(.bluetooth) }
        if core.contains(.persistence) { value.insert(.persistence) }
        if core.contains(.secureStorage) { value.insert(.secureStorage) }
        if core.contains(.networkTransfer) { value.insert(.networkTransfer) }
        if core.contains(.recordingSink) { value.insert(.recordingSink) }
        if core.contains(.firmwareBlob) { value.insert(.firmwareBlob) }
        self = value
    }
}

public struct DeviceReconnectHint: Equatable, Sendable {
    public var storedPeripheralID: String?
    public var advertisedAddress: String?
    public var storedName: String?
    public var scanTimeoutMilliseconds: UInt64
    public var connectionTimeoutMilliseconds: UInt64

    public init(
        storedPeripheralID: String? = nil,
        advertisedAddress: String? = nil,
        storedName: String? = nil,
        scanTimeoutMilliseconds: UInt64 = 10_000,
        connectionTimeoutMilliseconds: UInt64 = 10_000
    ) {
        self.storedPeripheralID = storedPeripheralID
        self.advertisedAddress = advertisedAddress
        self.storedName = storedName
        self.scanTimeoutMilliseconds = scanTimeoutMilliseconds
        self.connectionTimeoutMilliseconds = connectionTimeoutMilliseconds
    }
}

protocol CoreWorkflowRunning: Sendable {
    func run(
        _ command: CoreCommand,
        capabilities: CoreCapabilities
    ) async -> AsyncThrowingStream<CoreNotification, Error>
    func cancel(_ id: UUID) async throws
}

extension CoreEngineActor: CoreWorkflowRunning {}

struct DeviceRuntime: Sendable {
    let engine: any CoreWorkflowRunning
    let capabilities: CoreCapabilities
    let connection: DeviceConnectionRegistry
    let operations: DeviceOperationCoordinator
    let disconnect: @Sendable (String) async throws -> Void
    let readStatus: @Sendable (String) async throws -> DeviceStatus
    let statusUpdates: @Sendable (String) async throws -> AsyncThrowingStream<DeviceStatus, Error>
    let stopStatusUpdates: @Sendable (String) async throws -> Void
    let directWrite: @Sendable (String, String, String, Data) async throws -> Void
    let serializeConnectionSettings: @Sendable (DeviceConnectionSettings, DeviceType) throws -> Data
    let encodeDeviceCommand: @Sendable (UInt8) throws -> Data
    let registerProvisioning: @Sendable (String, @escaping ProvisioningMaterialProvider) async -> Void
    let registerFactoryReset: @Sendable (String, @escaping FactoryResetMaterialProvider) async -> Void
    let unregisterMaterial: @Sendable (String) async -> Void
    let registerFactoryResetGeneration: @Sendable (String, UInt64) async -> Void
    let unregisterFactoryResetGeneration: @Sendable (String) async -> Void
    let loadPendingFactoryReset: @Sendable () async throws -> PersistedFactoryResetResult?

    init(
        engine: any CoreWorkflowRunning,
        capabilities: CoreCapabilities,
        connection: DeviceConnectionRegistry = DeviceConnectionRegistry(),
        operations: DeviceOperationCoordinator = DeviceOperationCoordinator(),
        disconnect: @escaping @Sendable (String) async throws -> Void,
        readStatus: @escaping @Sendable (String) async throws -> DeviceStatus = { _ in
            throw NativeHostError.missingResource("device status")
        },
        statusUpdates: @escaping @Sendable (String) async throws -> AsyncThrowingStream<DeviceStatus, Error> = { _ in
            throw NativeHostError.missingResource("device status subscription")
        },
        stopStatusUpdates: @escaping @Sendable (String) async throws -> Void = { _ in },
        directWrite: @escaping @Sendable (String, String, String, Data) async throws -> Void = { _, _, _, _ in
            throw NativeHostError.missingResource("direct device write")
        },
        serializeConnectionSettings: @escaping @Sendable (DeviceConnectionSettings, DeviceType) throws -> Data = { _, _ in
            throw NativeHostError.missingResource("connection-settings encoder")
        },
        encodeDeviceCommand: @escaping @Sendable (UInt8) throws -> Data = { _ in
            throw NativeHostError.missingResource("device-command encoder")
        },
        registerProvisioning: @escaping @Sendable (String, @escaping ProvisioningMaterialProvider) async -> Void = { _, _ in },
        registerFactoryReset: @escaping @Sendable (String, @escaping FactoryResetMaterialProvider) async -> Void = { _, _ in },
        unregisterMaterial: @escaping @Sendable (String) async -> Void = { _ in },
        registerFactoryResetGeneration: @escaping @Sendable (String, UInt64) async -> Void = { _, _ in },
        unregisterFactoryResetGeneration: @escaping @Sendable (String) async -> Void = { _ in },
        loadPendingFactoryReset: @escaping @Sendable () async throws -> PersistedFactoryResetResult? = { nil }
    ) {
        self.engine = engine
        self.capabilities = capabilities
        self.connection = connection
        self.operations = operations
        self.disconnect = disconnect
        self.readStatus = readStatus
        self.statusUpdates = statusUpdates
        self.stopStatusUpdates = stopStatusUpdates
        self.directWrite = directWrite
        self.serializeConnectionSettings = serializeConnectionSettings
        self.encodeDeviceCommand = encodeDeviceCommand
        self.registerProvisioning = registerProvisioning
        self.registerFactoryReset = registerFactoryReset
        self.unregisterMaterial = unregisterMaterial
        self.registerFactoryResetGeneration = registerFactoryResetGeneration
        self.unregisterFactoryResetGeneration = unregisterFactoryResetGeneration
        self.loadPendingFactoryReset = loadPendingFactoryReset
    }
}

extension CoreCapabilities {
    static let all: Self = [
        .bluetooth, .timer, .persistence, .secureStorage, .networkTransfer,
        .progress, .hostMaterial, .recordingSink, .firmwareBlob,
    ]
}

public actor DeviceManager {
    private struct ActiveOperation {
        let cancellationID: UUID
        var task: Task<Void, Never>?
    }

    private struct StatusObserver {
        let peripheralID: String
        let continuation: AsyncThrowingStream<DeviceStatus, Error>.Continuation
        var task: Task<Void, Never>?
    }

    private var runtime: DeviceRuntime?
    private var activeOperation: ActiveOperation?
    private var connectedDevice: ConnectedDevice?
    private var connectionObservers: [UUID: AsyncStream<ConnectedDevice?>.Continuation] = [:]
    private var statusObservers: [UUID: StatusObserver] = [:]

    public init() {}

    func attach(_ runtime: DeviceRuntime) {
        self.runtime = runtime
    }

    func detach() async {
        if let activeOperation {
            activeOperation.task?.cancel()
            try? await runtime?.engine.cancel(activeOperation.cancellationID)
            await runtime?.operations.end(activeOperation.cancellationID)
        }
        await stopAllStatusObservers()
        if let connectedDevice { try? await runtime?.disconnect(connectedDevice.id) }
        await runtime?.connection.clear()
        activeOperation = nil
        connectedDevice = nil
        runtime = nil
        connectionObservers.values.forEach { $0.finish() }
        connectionObservers.removeAll()
    }

    public func capabilities() throws -> DeviceCapabilities {
        DeviceCapabilities(core: try configuredRuntime().capabilities)
    }

    public func startScan(
        timeoutMilliseconds: UInt64 = 10_000,
        allowDuplicates: Bool = false
    ) async throws -> AsyncThrowingStream<DiscoveredDevice, Error> {
        let runtime = try await beginOperation(operation: .discover)
        let cancellationID = activeOperation!.cancellationID
        let command = CoreCommand.discoverDevices(
            timeoutMilliseconds: timeoutMilliseconds,
            allowDuplicates: allowDuplicates,
            cancellationID: cancellationID
        )
        let pair = AsyncThrowingStream<DiscoveredDevice, Error>.makeStream()
        let task = Task {
            do {
                let notifications = await runtime.engine.run(command, capabilities: runtime.capabilities)
                for try await notification in notifications {
                    switch notification.kind {
                    case .deviceDiscovered:
                        pair.continuation.yield(try Self.discoveredDevice(notification))
                    case .failed:
                        throw Self.publicError(notification)
                    case .started, .connectionEstablished, .progress, .retrying,
                         .deviceUploadPreserved, .bleFallbackReady, .firmwareProgress,
                         .deviceLog, .completed, .cancelled:
                        break
                    }
                }
                pair.continuation.finish()
            } catch is CancellationError {
                pair.continuation.finish()
            } catch {
                pair.continuation.finish(throwing: Self.publicError(error))
            }
            await finishOperation(cancellationID)
        }
        activeOperation?.task = task
        pair.continuation.onTermination = { @Sendable _ in
            Task { await self.cancelIfActive(cancellationID) }
        }
        return pair.stream
    }

    public func connect(serialNumber: String, device: DiscoveredDevice) async throws -> ConnectedDevice {
        if let connectedDevice, connectedDevice.id != device.id {
            let runtime = try configuredRuntime()
            await stopAllStatusObservers()
            try await runtime.disconnect(connectedDevice.id)
            await runtime.connection.clear()
            self.connectedDevice = nil
            publishConnection()
        }
        return try await runConnection(
            .connect(
                serialNumber: serialNumber,
                peripheralID: device.id,
                name: device.name,
                advertisedAddress: device.macAddress,
                rssi: Int16(clamping: device.rssi)
            ),
            source: device
        )
    }

    public func reconnect(
        serialNumber: String,
        hint: DeviceReconnectHint = .init()
    ) async throws -> ConnectedDevice {
        try await runConnection(
            .reconnect(
                serialNumber: serialNumber,
                storedPeripheralID: hint.storedPeripheralID,
                advertisedAddress: hint.advertisedAddress,
                storedName: hint.storedName,
                scanTimeoutMilliseconds: hint.scanTimeoutMilliseconds,
                connectionTimeoutMilliseconds: hint.connectionTimeoutMilliseconds
            ),
            source: nil
        )
    }

    public func disconnect() async throws {
        guard let connectedDevice else { return }
        let runtime = try configuredRuntime()
        await stopAllStatusObservers()
        try await runtime.disconnect(connectedDevice.id)
        await runtime.connection.clear()
        self.connectedDevice = nil
        publishConnection()
    }

    public func cancelCurrentOperation() async throws {
        guard let activeOperation else { return }
        let runtime = try configuredRuntime()
        activeOperation.task?.cancel()
        try await runtime.engine.cancel(activeOperation.cancellationID)
        self.activeOperation = nil
        await runtime.operations.end(activeOperation.cancellationID)
    }

    public func connectionUpdates() -> AsyncStream<ConnectedDevice?> {
        let id = UUID()
        let pair = AsyncStream<ConnectedDevice?>.makeStream()
        connectionObservers[id] = pair.continuation
        pair.continuation.yield(connectedDevice)
        pair.continuation.onTermination = { @Sendable _ in
            Task { await self.removeConnectionObserver(id) }
        }
        return pair.stream
    }

    public func readStatus() async throws -> DeviceStatus {
        guard let connectedDevice else {
            throw BotaDeviceSDKError(
                code: .notConnected,
                operation: .readStatus,
                retryable: true,
                detail: "a verified device connection is required"
            )
        }
        return try await configuredRuntime().readStatus(connectedDevice.id)
    }

    public func statusUpdates() throws -> AsyncThrowingStream<DeviceStatus, Error> {
        guard let connectedDevice else {
            throw BotaDeviceSDKError(
                code: .notConnected,
                operation: .readStatus,
                retryable: true,
                detail: "a verified device connection is required"
            )
        }
        let runtime = try configuredRuntime()
        let id = UUID()
        let pair = AsyncThrowingStream<DeviceStatus, Error>.makeStream()
        statusObservers[id] = StatusObserver(
            peripheralID: connectedDevice.id,
            continuation: pair.continuation,
            task: nil
        )
        let task = Task {
            do {
                let updates = try await runtime.statusUpdates(connectedDevice.id)
                for try await status in updates { pair.continuation.yield(status) }
                pair.continuation.finish()
            } catch is CancellationError {
                pair.continuation.finish()
            } catch {
                pair.continuation.finish(throwing: Self.publicError(error))
            }
            statusObserverCompleted(id)
        }
        statusObservers[id]?.task = task
        pair.continuation.onTermination = { @Sendable _ in Task { await self.stopStatusObserver(id) } }
        return pair.stream
    }

    private func runConnection(_ command: CoreCommand, source: DiscoveredDevice?) async throws -> ConnectedDevice {
        let runtime = try await beginOperation(
            cancellationID: command.cancellationID,
            operation: command.kind == UInt32(BOTA_DEVICE_SDK_V1_COMMAND_RECONNECT) ? .reconnect : .connect
        )
        var established: ConnectedDevice?
        do {
            let notifications = await runtime.engine.run(command, capabilities: runtime.capabilities)
            for try await notification in notifications {
                switch notification.kind {
                case .connectionEstablished:
                    established = try Self.connectedDevice(notification, source: source)
                case .failed:
                    throw Self.publicError(notification)
                case .started, .deviceDiscovered, .progress, .retrying,
                     .deviceUploadPreserved, .bleFallbackReady, .firmwareProgress,
                     .deviceLog, .completed, .cancelled:
                    break
                }
            }
        } catch {
            await finishOperation(command.cancellationID)
            throw Self.publicError(error)
        }
        guard let established else {
            await finishOperation(command.cancellationID)
            throw BotaDeviceSDKError(
                code: .connectionFailed,
                operation: command.kind == UInt32(BOTA_DEVICE_SDK_V1_COMMAND_RECONNECT) ? .reconnect : .connect,
                retryable: true,
                detail: "connection completed without verified identity"
            )
        }
        connectedDevice = established
        await runtime.connection.set(established)
        publishConnection()
        await finishOperation(command.cancellationID)
        return established
    }

    private func configuredRuntime() throws -> DeviceRuntime {
        guard let runtime else {
            throw BotaDeviceSDKError(
                code: .featureUnavailable,
                operation: .validate,
                retryable: false,
                detail: "BotaDeviceClient.configure() must be called first"
            )
        }
        return runtime
    }

    private func beginOperation(
        cancellationID: UUID = UUID(),
        operation: BotaOperation
    ) async throws -> DeviceRuntime {
        let runtime = try configuredRuntime()
        guard activeOperation == nil else {
            throw BotaDeviceSDKError(
                code: .operationInProgress,
                operation: .validate,
                retryable: false,
                detail: "another device workflow is already active"
            )
        }
        try await runtime.operations.begin(cancellationID, operation: operation)
        activeOperation = ActiveOperation(cancellationID: cancellationID)
        return runtime
    }

    private func finishOperation(_ cancellationID: UUID) async {
        if activeOperation?.cancellationID == cancellationID {
            activeOperation = nil
            await runtime?.operations.end(cancellationID)
        }
    }

    private func cancelIfActive(_ cancellationID: UUID) async {
        guard activeOperation?.cancellationID == cancellationID else { return }
        activeOperation?.task?.cancel()
        try? await runtime?.engine.cancel(cancellationID)
        activeOperation = nil
        await runtime?.operations.end(cancellationID)
    }

    private func publishConnection() {
        connectionObservers.values.forEach { $0.yield(connectedDevice) }
    }

    private func removeConnectionObserver(_ id: UUID) { connectionObservers[id] = nil }
    private func statusObserverCompleted(_ id: UUID) { statusObservers[id] = nil }

    private func stopStatusObserver(_ id: UUID) async {
        guard let observer = statusObservers.removeValue(forKey: id) else { return }
        observer.task?.cancel()
        observer.continuation.finish()
        try? await runtime?.stopStatusUpdates(observer.peripheralID)
    }

    private func stopAllStatusObservers() async {
        let observers = statusObservers
        statusObservers.removeAll()
        for observer in observers.values {
            observer.task?.cancel()
            observer.continuation.finish()
            try? await runtime?.stopStatusUpdates(observer.peripheralID)
        }
    }

    private static func discoveredDevice(_ notification: CoreNotification) throws -> DiscoveredDevice {
        DiscoveredDevice(
            id: try text(notification, UInt32(BOTA_DEVICE_SDK_V1_FIELD_PERIPHERAL_ID)),
            name: optionalText(notification, UInt32(BOTA_DEVICE_SDK_V1_FIELD_NAME)),
            macAddress: optionalText(notification, UInt32(BOTA_DEVICE_SDK_V1_FIELD_ADVERTISED_ADDRESS)),
            rssi: Int(try signed(notification, UInt32(BOTA_DEVICE_SDK_V1_FIELD_RSSI)))
        )
    }

    private static func connectedDevice(
        _ notification: CoreNotification,
        source: DiscoveredDevice?
    ) throws -> ConnectedDevice {
        ConnectedDevice(
            id: try text(notification, UInt32(BOTA_DEVICE_SDK_V1_FIELD_PERIPHERAL_ID)),
            serialNumber: try text(notification, UInt32(BOTA_DEVICE_SDK_V1_FIELD_SERIAL_NUMBER)),
            deviceType: source?.deviceType ?? .unknown(0),
            firmwareVersion: source?.firmwareVersion ?? "",
            isProvisioned: false,
            connectionState: .connected,
            mtu: 0
        )
    }

    private static func publicError(_ notification: CoreNotification) -> BotaDeviceSDKError {
        let errorCode = notification.packet.fields.unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_ERROR_CODE)) ?? 21
        return BotaDeviceSDKError(
            code: BotaDeviceSDKError.code(UInt32(clamping: errorCode)),
            operation: BotaDeviceSDKError.operation(notification.packet.operation),
            retryable: notification.packet.fields.bool(UInt32(BOTA_DEVICE_SDK_V1_FIELD_RETRYABLE)) ?? false,
            protocolStatus: notification.packet.fields
                .unsigned(UInt32(BOTA_DEVICE_SDK_V1_FIELD_PROTOCOL_STATUS))
                .flatMap(UInt16.init(exactly:)),
            detail: optionalText(notification, UInt32(BOTA_DEVICE_SDK_V1_FIELD_ERROR_DETAIL)) ?? "device workflow failed"
        )
    }

    private static func publicError(_ error: Error) -> Error {
        if let error = error as? BotaDeviceSDKError { return error }
        if let error = error as? CoreError { return BotaDeviceSDKError(error) }
        return error
    }

    private static func text(_ notification: CoreNotification, _ id: UInt32) throws -> String {
        guard let value = optionalText(notification, id) else { throw NativeHostError.missingField(id) }
        return value
    }

    private static func optionalText(_ notification: CoreNotification, _ id: UInt32) -> String? {
        notification.packet.fields.text(id)
    }

    private static func signed(_ notification: CoreNotification, _ id: UInt32) throws -> Int64 {
        for field in notification.packet.fields {
            if case let .signed(fieldID, value) = field, fieldID == id { return value }
        }
        throw NativeHostError.missingField(id)
    }
}

private extension Array where Element == CoreField {
    func bool(_ id: UInt32) -> Bool? {
        for field in self {
            if case let .bool(fieldID, value) = field, fieldID == id { return value }
        }
        return nil
    }

    func text(_ id: UInt32) -> String? {
        for field in self {
            if case let .text(fieldID, value) = field, fieldID == id { return value }
        }
        return nil
    }
}

File: platforms/android/sdk/src/main/kotlin/dev/bota/sdk/DeviceManager.kt
package dev.bota.sdk

import dev.bota.sdk.internal.DeviceRuntime
import dev.bota.sdk.internal.core.CoreCapabilities
import dev.bota.sdk.internal.core.CoreCommand
import dev.bota.sdk.internal.core.CoreNotification
import dev.bota.sdk.internal.core.CoreNotificationKind
import dev.bota.sdk.internal.jni.NativeCoreException
import dev.bota.sdk.internal.toPublicError
import dev.bota.sdk.internal.workflowError
import dev.bota.sdk.model.ConnectedDevice
import dev.bota.sdk.model.ConnectionState
import dev.bota.sdk.model.DeviceStatus
import dev.bota.sdk.model.DeviceType
import dev.bota.sdk.model.DiscoveredDevice
import java.util.UUID
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.SendChannel
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

public enum class DeviceCapability {
    Bluetooth,
    Persistence,
    SecureStorage,
    NetworkTransfer,
    RecordingSink,
    FirmwareBlob,
}

public class DeviceCapabilities internal constructor(
    private val values: Set<DeviceCapability>,
) {
    public operator fun contains(capability: DeviceCapability): Boolean = capability in values
    public fun asSet(): Set<DeviceCapability> = values.toSet()
}

public data class DeviceReconnectHint(
    public val storedPeripheralId: String? = null,
    public val advertisedAddress: String? = null,
    public val storedName: String? = null,
    public val scanTimeoutMilliseconds: ULong = 10_000u,
    public val connectionTimeoutMilliseconds: ULong = 10_000u,
)

public class DeviceManager internal constructor() {
    private data class ActiveOperation(val cancellationId: UUID, var task: Job? = null)
    private data class StatusObserver(val peripheralId: String, val channel: SendChannel<DeviceStatus>, val task: Job)

    private val lock = Any()
    private var callbackScope: CoroutineScope? = null
    private var runtime: DeviceRuntime? = null
    private var activeOperation: ActiveOperation? = null
    private var connectedDevice: ConnectedDevice? = null
    private val connectionObservers = mutableMapOf<UUID, SendChannel<ConnectedDevice?>>()
    private val statusObservers = mutableMapOf<UUID, StatusObserver>()

    internal fun attach(runtime: DeviceRuntime) {
        synchronized(lock) {
            this.runtime = runtime
            callbackScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
        }
    }

    internal suspend fun detach() {
        val snapshot = synchronized(lock) {
            val value = DetachSnapshot(
                runtime = runtime,
                active = activeOperation,
                connected = connectedDevice,
                status = statusObservers.values.toList(),
                connection = connectionObservers.values.toList(),
                callbackScope = callbackScope,
            )
            runtime = null
            activeOperation = null
            connectedDevice = null
            statusObservers.clear()
            connectionObservers.clear()
            callbackScope = null
            value
        }
        snapshot.active?.task?.cancel()
        snapshot.active?.let { runCatching { snapshot.runtime?.engine?.cancel(it.cancellationId) } }
        snapshot.status.forEach { observer ->
            observer.task.cancel()
            observer.channel.close()
            runCatching { snapshot.runtime?.stopStatusUpdates?.invoke(observer.peripheralId) }
        }
        snapshot.connected?.let { device -> runCatching { snapshot.runtime?.disconnect?.invoke(device.id) } }
        snapshot.connection.forEach { it.close() }
        snapshot.callbackScope?.cancel()
    }

    public fun capabilities(): DeviceCapabilities = DeviceCapabilities(configuredRuntime().capabilities.publicValues())

    public suspend fun startScan(
        timeoutMilliseconds: ULong = 10_000u,
        allowDuplicates: Boolean = false,
    ): Flow<DiscoveredDevice> {
        val configured = configuredRuntime()
        configured.authorize(BotaOperation.Discover)
        val managerScope = configuredCallbackScope()
        beginOperation(BotaOperation.Discover)
        val operation = synchronized(lock) { requireNotNull(activeOperation) }
        val command = CoreCommand.discoverDevices(
            timeoutMilliseconds = timeoutMilliseconds,
            allowDuplicates = allowDuplicates,
            cancellationId = operation.cancellationId,
        )
        return callbackFlow {
            val task = launch(start = CoroutineStart.LAZY) {
                var cancelled = false
                try {
                    configured.engine.run(command, configured.capabilities).collect { notification ->
                        when (notification.kind) {
                            CoreNotificationKind.DeviceDiscovered -> send(notification.discoveredDevice())
                            CoreNotificationKind.Failed -> throw notification.workflowError()
                            CoreNotificationKind.Cancelled -> throw cancelled(BotaOperation.Discover)
                            else -> Unit
                        }
                    }
                    close()
                } catch (error: CancellationException) {
                    cancelled = true
                    throw error
                } catch (error: Throwable) {
                    close(error.toPublicError())
                } finally {
                    withContext(NonCancellable) {
                        if (cancelled) cancelIfActive(operation.cancellationId)
                        else finishOperation(operation.cancellationId)
                    }
                }
            }
            synchronized(lock) {
                activeOperation?.takeIf { it.cancellationId == operation.cancellationId }?.task = task
            }
            task.start()
            awaitClose {
                task.cancel()
                managerScope.launch { cancelIfActive(operation.cancellationId) }
            }
        }
    }

    public suspend fun connect(serialNumber: String, device: DiscoveredDevice): ConnectedDevice {
        requireSerial(serialNumber, BotaOperation.Connect)
        return runConnection(
            command = CoreCommand.connect(
                serialNumber = serialNumber,
                peripheralId = device.id,
                name = device.name,
                advertisedAddress = device.macAddress,
                rssi = device.rssi,
            ),
            expectedSerial = serialNumber,
            source = device,
            operation = BotaOperation.Connect,
        )
    }

    public suspend fun reconnect(
        serialNumber: String,
        hint: DeviceReconnectHint = DeviceReconnectHint(),
    ): ConnectedDevice {
        requireSerial(serialNumber, BotaOperation.Reconnect)
        return runConnection(
            command = CoreCommand.reconnect(
                serialNumber = serialNumber,
                storedPeripheralId = hint.storedPeripheralId,
                advertisedAddress = hint.advertisedAddress,
                storedName = hint.storedName,
                scanTimeoutMilliseconds = hint.scanTimeoutMilliseconds,
                connectionTimeoutMilliseconds = hint.connectionTimeoutMilliseconds,
            ),
            expectedSerial = serialNumber,
            source = null,
            operation = BotaOperation.Reconnect,
        )
    }

    public suspend fun disconnect() {
        val configured = configuredRuntime()
        val device = synchronized(lock) { connectedDevice } ?: return
        stopAllStatusObservers()
        configured.authorize(BotaOperation.Connect)
        configured.disconnect(device.id)
        synchronized(lock) { if (connectedDevice?.id == device.id) connectedDevice = null }
        publishConnection()
    }

    public suspend fun cancelCurrentOperation() {
        val operation = synchronized(lock) {
            activeOperation.also { activeOperation = null }
        } ?: return
        operation.task?.cancel()
        configuredRuntime().engine.cancel(operation.cancellationId)
    }

    public fun connectionUpdates(): Flow<ConnectedDevice?> = callbackFlow {
        val id = UUID.randomUUID()
        val initial = synchronized(lock) {
            connectionObservers[id] = channel
            connectedDevice
        }
        send(initial)
        awaitClose { synchronized(lock) { connectionObservers.remove(id) } }
    }

    public suspend fun readStatus(): DeviceStatus {
        val configured = configuredRuntime()
        val device = requireConnected()
        configured.authorize(BotaOperation.ReadStatus)
        return configured.decodeStatus(configured.readStatus(device.id))
    }

    public fun statusUpdates(): Flow<DeviceStatus> {
        val configured = configuredRuntime()
        val device = requireConnected()
        configured.authorize(BotaOperation.ReadStatus)
        val managerScope = configuredCallbackScope()
        return callbackFlow {
            val id = UUID.randomUUID()
            val task = launch(start = CoroutineStart.LAZY) {
                var failure: Throwable? = null
                try {
                    configured.statusUpdates(device.id).collect { send(configured.decodeStatus(it)) }
                } catch (error: CancellationException) {
                    throw error
                } catch (error: Throwable) {
                    failure = error.toPublicError()
                } finally {
                    withContext(NonCancellable) { stopStatusObserver(id, cancelTask = false) }
                }
                failure?.let(::close) ?: close()
            }
            synchronized(lock) { statusObservers[id] = StatusObserver(device.id, channel, task) }
            task.start()
            awaitClose {
                task.cancel()
                managerScope.launch { stopStatusObserver(id, cancelTask = true) }
            }
        }
    }

    private suspend fun runConnection(
        command: CoreCommand,
        expectedSerial: String,
        source: DiscoveredDevice?,
        operation: BotaOperation,
    ): ConnectedDevice {
        val configured = configuredRuntime()
        configured.authorize(operation)
        beginOperation(operation, command.cancellationId)
        var established: ConnectedDevice? = null
        try {
            source?.let { disconnectDifferentDevice(it.id) }
            configured.engine.run(command, configured.capabilities).collect { notification ->
                when (notification.kind) {
                    CoreNotificationKind.ConnectionEstablished -> {
                        val candidate = notification.connectedDevice(source)
                        if (candidate.serialNumber != expectedSerial) {
                            throw identityMismatch(expectedSerial, operation)
                        }
                        established = candidate
                    }
                    CoreNotificationKind.Failed -> throw notification.workflowError()
                    CoreNotificationKind.Cancelled -> throw cancelled(operation)
                    else -> Unit
                }
            }
        } catch (error: CancellationException) {
            withContext(NonCancellable) { cancelIfActive(command.cancellationId) }
            throw error
        } catch (error: Throwable) {
            finishOperation(command.cancellationId)
            throw error.toPublicError()
        }
        val connected = established ?: run {
            finishOperation(command.cancellationId)
            throw coreError(
                BotaErrorCode.ConnectionFailed,
                operation,
                retryable = true,
                "connection completed without verified identity",
            )
        }
        synchronized(lock) { connectedDevice = connected }
        publishConnection()
        finishOperation(command.cancellationId)
        return connected
    }

    private fun configuredRuntime(): DeviceRuntime = synchronized(lock) { runtime } ?: throw coreError(
        BotaErrorCode.FeatureUnavailable,
        BotaOperation.Validate,
        retryable = false,
        "BotaDeviceClient.configure() must be called first",
    )

    private fun beginOperation(operation: BotaOperation, cancellationId: UUID = UUID.randomUUID()): DeviceRuntime {
        val configured = configuredRuntime()
        synchronized(lock) {
            if (activeOperation != null) throw coreError(
                BotaErrorCode.OperationInProgress,
                operation,
                retryable = false,
                "another device workflow is already active",
            )
            activeOperation = ActiveOperation(cancellationId)
        }
        return configured
    }

    private fun finishOperation(cancellationId: UUID) {
        synchronized(lock) {
            if (activeOperation?.cancellationId == cancellationId) activeOperation = null
        }
    }

    private suspend fun cancelIfActive(cancellationId: UUID) {
        val configured = synchronized(lock) {
            if (activeOperation?.cancellationId != cancellationId) return
            activeOperation = null
            runtime
        } ?: return
        runCatching { configured.engine.cancel(cancellationId) }
    }

    private suspend fun disconnectDifferentDevice(nextId: String) {
        val configured = configuredRuntime()
        val current = synchronized(lock) { connectedDevice } ?: return
        if (current.id == nextId) return
        stopAllStatusObservers()
        configured.authorize(BotaOperation.Connect)
        configured.disconnect(current.id)
        synchronized(lock) { if (connectedDevice?.id == current.id) connectedDevice = null }
        publishConnection()
    }

    private fun requireConnected(): ConnectedDevice = synchronized(lock) { connectedDevice } ?: throw coreError(
        BotaErrorCode.NotConnected,
        BotaOperation.ReadStatus,
        retryable = true,
        "a verified device connection is required",
    )

    private fun publishConnection() {
        val (device, observers) = synchronized(lock) { connectedDevice to connectionObservers.values.toList() }
        observers.forEach { it.trySend(device) }
    }

    private suspend fun stopStatusObserver(id: UUID, cancelTask: Boolean) {
        val observer = synchronized(lock) { statusObservers.remove(id) } ?: return
        if (cancelTask) observer.task.cancel()
        observer.channel.close()
        runCatching { runtime?.stopStatusUpdates?.invoke(observer.peripheralId) }
    }

    private suspend fun stopAllStatusObservers() {
        val observers = synchronized(lock) { statusObservers.values.toList().also { statusObservers.clear() } }
        observers.forEach {
            it.task.cancel()
            it.channel.close()
            runCatching { runtime?.stopStatusUpdates?.invoke(it.peripheralId) }
        }
    }

    private fun configuredCallbackScope(): CoroutineScope = synchronized(lock) { callbackScope }
        ?: throw coreError(
            BotaErrorCode.FeatureUnavailable,
            BotaOperation.Validate,
            retryable = false,
            "BotaDeviceClient.configure() must be called first",
        )

    private data class DetachSnapshot(
        val runtime: DeviceRuntime?,
        val active: ActiveOperation?,
        val connected: ConnectedDevice?,
        val status: List<StatusObserver>,
        val connection: List<SendChannel<ConnectedDevice?>>,
        val callbackScope: CoroutineScope?,
    )
}

private fun CoreNotification.discoveredDevice(): DiscoveredDevice = DiscoveredDevice(
    id = requiredText(4),
    name = packet.texts(5).firstOrNull(),
    macAddress = packet.texts(6).firstOrNull(),
    rssi = packet.signed(7)?.toInt() ?: missingField(7),
)

private fun CoreNotification.connectedDevice(source: DiscoveredDevice?): ConnectedDevice = ConnectedDevice(
    id = requiredText(4),
    serialNumber = requiredText(3),
    deviceType = source?.deviceType ?: DeviceType.Unknown(0u),
    firmwareVersion = source?.firmwareVersion ?: "",
    isProvisioned = false,
    connectionState = ConnectionState.Connected,
    mtu = packet.unsigneds(57).firstOrNull()?.toInt() ?: 0,
)

private fun CoreNotification.requiredText(id: Int): String = packet.texts(id).firstOrNull() ?: missingField(id)

private fun missingField(id: Int): Nothing = throw coreError(
    BotaErrorCode.UnexpectedEvent,
    BotaOperation.Decode,
    retryable = false,
    "workflow notification is missing field $id",
)

private fun requireSerial(value: String, operation: BotaOperation) {
    if (value.isBlank()) throw coreError(
        BotaErrorCode.InvalidInput,
        operation,
        retryable = false,
        "serial number is required",
    )
}

private fun identityMismatch(serialNumber: String, operation: BotaOperation): BotaSDKError.Core = coreError(
    BotaErrorCode.IdentityMismatch,
    operation,
    retryable = false,
    "connected device did not verify serial number $serialNumber",
)

private fun cancelled(operation: BotaOperation): BotaSDKError.Core = coreError(
    BotaErrorCode.Cancelled,
    operation,
    retryable = true,
    "device workflow was cancelled",
)

private fun coreError(
    code: BotaErrorCode,
    operation: BotaOperation,
    retryable: Boolean,
    detail: String,
): BotaSDKError.Core = BotaSDKError.Core(code, operation, retryable, null, detail)

private fun Throwable.toPublicError(): Throwable = when (this) {
    is BotaSDKError -> this
    is NativeCoreException -> toPublicError()
    else -> this
}

private fun CoreCapabilities.publicValues(): Set<DeviceCapability> = buildSet {
    if (bits and CoreCapabilities.Bluetooth.bits != 0uL) add(DeviceCapability.Bluetooth)
    if (bits and CoreCapabilities.Persistence.bits != 0uL) add(DeviceCapability.Persistence)
    if (bits and CoreCapabilities.SecureStorage.bits != 0uL) add(DeviceCapability.SecureStorage)
    if (bits and CoreCapabilities.NetworkTransfer.bits != 0uL) add(DeviceCapability.NetworkTransfer)
    if (bits and CoreCapabilities.RecordingSink.bits != 0uL) add(DeviceCapability.RecordingSink)
    if (bits and CoreCapabilities.FirmwareBlob.bits != 0uL) add(DeviceCapability.FirmwareBlob)
}
